src/Lean/Compiler/IR/LLVMBindings.lean

```diff
@@ -11,282 +14,303 @@ are to:
   data structures to build IR.
 -/
 
-namespace LLVM
-
-namespace CodeGenFileType
 -- https://github.com/llvm/llvm-project/blob/c3e073bcbdc523b0f758d44a89a6333e38bff863/llvm/include/llvm-c/TargetMachine.h#L64
-def AssemblyFile : UInt64 := 0
-def ObjectFile : UInt64 := 1
-end CodeGenFileType
+structure CodegenFileType where
+  private mk :: val : UInt64
+
+def CodegenFileType.AssemblyFile : CodegenFileType := { val := 0 }
+def CodegenFileType.ObjectFile : CodegenFileType := { val := 1 }
 
-namespace IntPredicate
 -- https://github.com/llvm/llvm-project/blob/c3e073bcbdc523b0f758d44a89a6333e38bff863/llvm/include/llvm-c/Core.h#L290
-def EQ : UInt64 := 32
-def NE : UInt64 := EQ + 1
-def UGT : UInt64 := NE + 1
-end IntPredicate
-
-opaque BasicBlock : Type
-opaque Context : Type
-opaque Module : Type
-opaque Builder : Type
-opaque LLVMType : Type
-opaque Value : Type
-opaque MemoryBuffer : Type
-opaque Target : Type
-opaque TargetMachine : Type
-
--- A raw pointer to a C object, whose Lean representation
--- is given by α
-structure Ptr (α: Type): Type where
-  ptr: USize
+structure IntPredicate where
+  private mk :: val : UInt64
+
+def IntPredicate.EQ : IntPredicate := { val := 32 }
+def IntPredicate.NE : IntPredicate := { val := IntPredicate.EQ.val + 1 }
+def IntPredicate.UGT : IntPredicate := { val := IntPredicate.NE.val + 1 }
+
+structure BasicBlock (ctx : Context)  where
+  private mk :: ptr : USize
+  deriving Inhabited
```

Comment:
Hmm, using this instance is unsafe I suppose, but I don't have a better solution except for transforming affected `opaque`s into `axiom`s

Reply:
You can turn it into a `Nonempty (BasicBlock ctx)` instance, which is safe.